lib/secrets/secrets.go

```diff
@@ -229,3 +279,155 @@ func PrepareSSHPrivateKey(keyContent string) (string, error) {
 	}
 	return keyPath.Name(), nil
 }
+
+// checks if encryption key exists or not
+func keyExists(basePath string) (bool, error) {
+	path := filepath.Join(basePath, "key")
+	return fileutils.PathExists(path)
+}
+
+func encodeBase64(b []byte) string {
+	return base64.StdEncoding.EncodeToString(b)
+}
+
+func decodeBase64(s string) ([]byte, error) {
+	data, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode base64 string")
+	}
+	return data, nil
+}
+
+// Takes in key and the data, then encrypts the base64 encoded data
+// Uses AES256 CTR mode for encryption
+// returns the ciphertext, iv and error
+func encrypt(key []byte, value []byte) ([]byte, []byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error creating cipher block during encryption")
+	}
+	iv, err := GenerateRandomBytes(block.BlockSize())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not generate random iv for encryption")
+	}
+	b := encodeBase64(value)
+	stream := cipher.NewCTR(block, iv)
+	ciphertext := make([]byte, len(b))
+	stream.XORKeyStream(ciphertext, []byte(b))
+	return ciphertext, iv, nil
+}
+
+// decrpting the cipher for ctr mode
+// returns the decoded data ( originally encoded with base64 ) and error
+func decrypt(block cipher.Block, ciphertext []byte, iv []byte) ([]byte, error) {
+	stream := cipher.NewCTR(block, iv)
+	plain := make([]byte, len(ciphertext))
+	stream.XORKeyStream(plain, ciphertext)
+	d, err := decodeBase64(string(plain))
+	if err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+// function used to write the secret values to given file in toml format
+// stores the name of the algorithm , iv and the ciphertext
+func writeToml(algorithm string, iv string, ciphertext string, path string) error {
+	secretData := SecretKeyToml{
+		Algorithm:  algorithm,
+		IV:         iv,
+		Ciphertext: ciphertext,
+	}
+
+	buf := new(bytes.Buffer)
+	if err := toml.NewEncoder(buf).Encode(secretData); err != nil {
+		log.Fatal(err)
+	}
+
+	err := fileutils.AtomicWrite(path, bytes.NewReader(buf.Bytes()), fileutils.WithAtomicWriteFileMode(0700))
```

Comment:
This file doesn't need execute permissions. Also, i think you need to set the owner of the file

Reply:
setting owner now